### PR TITLE
WIP: Element Builder proposal

### PR DESCRIPTION
**Pull request by tdurieux**
I created this PR in order to discuss the architecture of a possible new Element Builder in spoon.

The current factories are neither easy to use nor explicit. This new builder aims to solve these issues.
But I don't know yet the best solution.

This PR contains a POC of the builder. I would like to have your opinion.

# Syntaxe

## BinaryOperator
```java
Builder B = factory.Builder();
CtBinaryOperator binary = B.Binary(
		B.Literal(2))
		.plus(B.Literal(1))
		.lower(
				B.Binary(B.Literal(10))
						.minus(B.Literal(5))
		).build();
Assert.assertEquals("(2 + 1) < (10 - 5)", binary.toString());
```

## Try

```java
Builder B = factory.Builder();
CtTry aTry = B.Try()
		.inBody(B.Increment(x.Literal(1)))
		.createCatch()
			.parameter("e", IllegalArgumentException.class)
			.inBody(x.Increment(x.Literal(1)))
			.close()
		.createCatch()
			.parameter("e", Exception.class)
			.close()
		.inFinally(x.Decrement(x.Literal(1)))
.build();
Assert.assertEquals("try {\n"
		+ "    1++;\n"
		+ "} catch (java.lang.IllegalArgumentException e) {\n"
		+ "    1++;\n"
		+ "} catch (java.lang.Exception e) {\n"
		+ "} finally {\n"
		+ "    1--;\n"
		+ "}", aTry.toString());
```
## Method

```java
Builder x = factory.Builder();

		CtReturn<Object> aReturn = factory.Core().createReturn();
		aReturn.setReturnedExpression(x.Literal(1).build());
		CtMethod m = x.Method("m")
				.type(int.class)
				.setPrivate()
				.setStatic()
				.addParam(int.class, "x")
				.inBody(aReturn).build();
		Assert.assertEquals("private static int m(int x) {\n"
				+ "    return 1;\n"
				+ "}", m.toString());
```
# Limitations
1. The builder requires a method to indicate the end of the element (here `CatchBuilder.close()`) in order to get back the parent Builder (here `TryBuilder`). It is not intuitive... 

2. This architecture will introduce a lot of duplicate codes. Example: all elements that have a body will have the method `inBody`. Unfortunately the inheritance will not solve this issue (in this case we need multi inheritance)

### CODE REVIEW

**Comment by alcides** [comment]
Love it! I really need this!

**Comment by monperrus** [comment]
looks like javapoet for Spoon?

**Comment by tdurieux** [comment]
I don't know

**Comment by GerardPaligot** [comment]
@monperrus Yes but better, with love and for Spoon! <3

File: src/test/java/spoon/reflect/builder/BinaryBuilderTest.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+		Builder B = factory.Builder();
+		CtUnaryOperator not = B.Not(B.Binary(B.Literal(2)).lower(B.Literal(1))).build();
+		Assert.assertEquals("!(2 < 1)", not.toString());
+	}
+
+	@Test
+	public void simpleBinaryTest() {
+		Factory factory = new Launcher().getFactory();
+		Builder B = factory.Builder();
+		CtBinaryOperator binary = B.Binary(
```
Comment (monperrus) [inline src/test/java/spoon/reflect/builder/BinaryBuilderTest.java]: infix notation? `Lit(2).times(Lit(3))`

File: src/test/java/spoon/reflect/builder/TryBuilderTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+public class TryBuilderTest {
+
+	@Test
+	public void test() {
+		Factory factory = new Launcher().getFactory();
+		Builder x = factory.Builder();
+
+CtTry aTry = x.Try()
+		.inBody(x.Increment(x.Literal(1)))
+		.createCatch()
```
Comment (monperrus) [inline src/test/java/spoon/reflect/builder/TryBuilderTest.java]: createCatch("e", IllegalArgumentException.class)

Comment (monperrus) [inline src/test/java/spoon/reflect/builder/TryBuilderTest.java]: .body(...)

**Comment by monperrus** [comment]
I love this idea. It will be an excellent alternative for intercession in addition to the core intercession API, snippets and templates.

**Comment by monperrus** [comment]
generate the builder with https://github.com/google/FreeBuilder?

**Comment by pvojtechovsky** [comment]
>The builder requires a method to indicate the end of the element (here CatchBuilder.close()) in order to get back the parent Builder (here TryBuilder). It is not intuitive...

What about API like this?
```java
Builder B = factory.Builder();
CtTry aTry = B.Try()
        .inBody(B.Increment(x.Literal(1)))
        .createCatch(B.Catch("e", IllegalArgumentException.class)
            .inBody(x.Increment(x.Literal(1)))
        ).createCatch(B.Catch("e", Exception.class)
        ).inFinally(x.Decrement(x.Literal(1)))
.build();
```
the close() can be avoided by building whole catch content as a parameter of createCatch(...)

**Comment by tdurieux** [comment]
Why not but there are a lot of parenthesis and call to the factory (that I was trying to avoid)

**Comment by pvojtechovsky** [comment]
```java
CtTry aTry = B.Try()
        .inBody(B.Increment(x.Literal(1)))
        .createCatch()
            .parameter("e", IllegalArgumentException.class)
            .inBody(x.Increment(x.Literal(1)))
            .close()
        .createCatch()
            .parameter("e", Exception.class)
            .close()
        .inFinally(x.Decrement(x.Literal(1)))
.build();

**Comment by tdurieux** [comment]
I did not understand your message sorry

**Comment by tdurieux** [comment]
> I do not prefer any yet. Just giving ideas into this nice concept, to help to move it forward 

Yes, it is very hard to create a good design for this builder.

> or just rename of methods?

I prefer this solution

Or use the build method

**Comment by pvojtechovsky** [comment]
> there are a lot of parenthesis and call to the factory (that I was trying to avoid)

May be we can avoid other calls of factory too
```java
CtTry aTry = B.Try()
        .beginTry()
            .Increment(x.Literal(1)))
        .endTry()
        .beginCatch()...
```
the `endTry()`, cannot be replaced by `end()`, because `endTry()` returns CtTryBuilder instance. The consistency of begin/end pairs would be assured in runtime by builder.

But I agree that this idea is really wild. So parenthesis and use of factory is better in this case, because we are building another spoon model element.

So the rule would be:
* building of each spoon model element starts on factory
* fluent API is used to fill all attributes of that one spoon model element

**Comment by tdurieux** [comment]
I would said:

CtTry aTry = B.Try()
        .beginBody()
            .Increment(x.Literal(1)))
        .endBody()
        .beginCatch()...
```

**Comment by pvojtechovsky** [comment]
> This architecture will introduce a lot of duplicate codes.

it is no problem, if code of builder will be generated.

**Comment by pvojtechovsky** [comment]
```java
CtTry aTry = B.Try()
.beginBody()
   .Increment(x.Literal(1)))
.endBody()
.beginCatch()
    .parameter("e", IllegalArgumentException.class)
    .beginBody()
       .Increment(x.Literal(1)))
     .endBody()
.endCatch()
```
The problem of this API is "What is the return type of `endBody() method"? Will be the same endBody method used in building of CtMethod?
A) If `endBody` returns generic `SpoonModelBuilder` class which is model element independent, then it there will be a lot of methods and API client will be not guided.
B) If `endBody` returns CtTry specific `CtTryBuilder` class, then how will be the `Increment(..)` method implemented to be able to return CtTryBuilder ??
Therefore I suggested `endTry` ... but I do not think that it is good way at all.

**Comment by tdurieux** [comment]
You are right, I did not think enough

**Comment by pvojtechovsky** [comment]
One more idea ... to use lambda expression to generate nested elements. Like this:
```java
CtTry aTry = factory.Builder().Try()
        .inBody(body ->
             body.Increment(x.Literal(1)))
        .createCatch(IllegalArgumentException.class, e, body->
             body.statement(st -> st.Increment(x.Literal(1)))
                     .statement(st -> st.invoke("java.lang.System.out.println()"))
        ).createCatch(Exception.class, "e")
        .inFinally(body -> 
             body.statemenet(st -> st.Decrement(x.Literal(1)))
.build();
```
* autocompletion will work on each level
* builder code has no duplicities

**Comment by tdurieux** [comment]
I can be a solution but it is only nice for Java 8.
In Java 7 it is not practicable but I don't know if it is a problem

**Pull request closed without merging**